File: python/test/t_MeshDomain2_std.py
#! /usr/bin/env python

import openturns as ot
import openturns.testing as ott
import otmeshing as otm

ot.TESTPREAMBLE()

# single-point API and sample API
for dim in range(2, 6):
    mesh = ot.IntervalMesher([1] * dim).build(ot.Interval([0.0] * dim, [1.0] * dim))
    domain = otm.MeshDomain2(mesh)

    # inside via single point
    p = [0.2] * dim
    distance = domain.computeDistance(p)
    ott.assert_almost_equal(distance, -0.2)

    # inside via sample
    ps = ot.Sample(1, p)
    distances = domain.computeDistance(ps)
    ott.assert_almost_equal(distances[0, 0], -0.2)

    # outside
    p = [1.2] * dim
    distance = domain.computeDistance(p)
    ott.assert_almost_equal(distance, 0.2 * dim**0.5)

    # far outside
    p = [10.0] * dim
    distance = domain.computeDistance(p)
    ott.assert_almost_equal(distance, 9.0 * dim**0.5)

    # point on the boundary
    p = [0.0] * dim
    distance = domain.computeDistance(p)
    ott.assert_almost_equal(distance, 0.0)

    # wrong dimension
    try:
        domain.computeDistance([0.0] * (dim + 1))
        assert False
    except RuntimeError:
        pass

# disjoint domain [0,1] U [2,3]
for dim in range(2, 6):
    mesh1 = ot.IntervalMesher([1] * dim).build(ot.Interval([0.0] * dim, [1.0] * dim))
    mesh2 = ot.IntervalMesher([1] * dim).build(ot.Interval([2.0] * dim, [3.0] * dim))
    mesh = otm.UnionMesher().build([mesh1, mesh2])
    domain = otm.MeshDomain2(mesh)

    # inside block 1
    p = [0.1] * dim
    distance = domain.computeDistance(p)
    ott.assert_almost_equal(distance, -0.1)

    # near block 1 (outside)
    p = [1.1] * dim
    distance = domain.computeDistance(p)
    ott.assert_almost_equal(distance, 0.1 * dim**0.5)

    # inside block 2
    p = [2.1] * dim
    distance = domain.computeDistance(p)
    ott.assert_almost_equal(distance, -0.1)

    # near block 2 (outside)
    p = [1.9] * dim
    distance = domain.computeDistance(p)
    ott.assert_almost_equal(distance, 0.1 * dim**0.5)

# 3D unit cube with known distances
mesh = ot.IntervalMesher([2] * 3).build(ot.Interval([0.0] * 3, [1.0] * 3))
domain = otm.MeshDomain2(mesh)

# center
distance = domain.computeDistance([0.5] * 3)
ott.assert_almost_equal(distance, -0.5)

# outside
distance = domain.computeDistance([1.5] * 3)
ott.assert_almost_equal(distance, 0.5 * 3.0**0.5)

# edge
distance = domain.computeDistance([0.5, 0.5, 0.0])
ott.assert_almost_equal(distance, 0.0)

File: lib/test/t_MeshDomain2_std.cxx
#include <iostream>
#include <cmath>

#include "otmeshing/otmeshing.hxx"
#include "otmeshing/MeshDomain2.hxx"

using namespace OT;
using namespace OTMESHING;

int main()
{
  // 2D unit square
  Sample vertices(4, 2);
  vertices(0, 0) = 0.0; vertices(0, 1) = 0.0;
  vertices(1, 0) = 1.0; vertices(1, 1) = 0.0;
  vertices(2, 0) = 1.0; vertices(2, 1) = 1.0;
  vertices(3, 0) = 0.0; vertices(3, 1) = 1.0;
  IndicesCollection simplices(2, 3);
  simplices(0, 0) = 0; simplices(0, 1) = 1; simplices(0, 2) = 2;
  simplices(1, 0) = 0; simplices(1, 1) = 2; simplices(1, 2) = 3;
  Mesh mesh2d(vertices, simplices);

  MeshDomain2 domain2d(mesh2d);
  std::cout << "domain2d=" << domain2d << std::endl;

  // inside point
  Point pInside2d({0.2, 0.2});
  Scalar d = domain2d.computeDistance(pInside2d);
  std::cout << "inside 2d distance=" << d << std::endl;
  if (std::abs(d + 0.2) > 1e-14)
    throw std::runtime_error("Expected distance -0.2");

  // outside point
  Point pOutside2d({1.2, 1.2});
  d = domain2d.computeDistance(pOutside2d);
  std::cout << "outside 2d distance=" << d << std::endl;
  if (std::abs(d - 0.2 * std::sqrt(2.0)) > 1e-14)
    throw std::runtime_error("Expected distance 0.2*sqrt(2)");

  // 3D unit cube
  Sample vertices3d(8, 3);
  for (UnsignedInteger i = 0; i < 8; ++ i)
    for (UnsignedInteger j = 0; j < 3; ++ j)
      vertices3d(i, j) = (i >> j) & 1;
  IndicesCollection simplices3d(6, 4);
  // tetrahedralization of a cube as 6 tets
  simplices3d(0, 0) = 0; simplices3d(0, 1) = 1; simplices3d(0, 2) = 3; simplices3d(0, 3) = 7;
  simplices3d(1, 0) = 0; simplices3d(1, 1) = 1; simplices3d(1, 2) = 5; simplices3d(1, 3) = 7;
  simplices3d(2, 0) = 0; simplices3d(2, 1) = 2; simplices3d(2, 2) = 3; simplices3d(2, 3) = 7;
  simplices3d(3, 0) = 0; simplices3d(3, 1) = 2; simplices3d(3, 2) = 6; simplices3d(3, 3) = 7;
  simplices3d(4, 0) = 0; simplices3d(4, 1) = 4; simplices3d(4, 2) = 5; simplices3d(4, 3) = 7;
  simplices3d(5, 0) = 0; simplices3d(5, 1) = 4; simplices3d(5, 2) = 6; simplices3d(5, 3) = 7;
  Mesh mesh3d(vertices3d, simplices3d);

  MeshDomain2 domain3d(mesh3d);
  std::cout << "domain3d=" << domain3d << std::endl;

  // inside point
  Point pInside3d({0.2, 0.2, 0.2});
  d = domain3d.computeDistance(pInside3d);
  std::cout << "inside 3d distance=" << d << std::endl;
  if (std::abs(d + 0.2) > 1e-14)
    throw std::runtime_error("Expected distance -0.2");

  // outside point
  Point pOutside3d({1.2, 1.2, 1.2});
  d = domain3d.computeDistance(pOutside3d);
  std::cout << "outside 3d distance=" << d << std::endl;
  if (std::abs(d - 0.2 * std::sqrt(3.0)) > 1e-14)
    throw std::runtime_error("Expected distance 0.2*sqrt(3)");

  std::cout << "ok" << std::endl;
  return 0;
}
